pkg/reconciler/buildlimitcleanup/build_limit_cleanup.go

```diff
@@ -0,0 +1,134 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package buildlimitcleanup
+
+import (
+	"context"
+	"sort"
+
+	build "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	"github.com/shipwright-io/build/pkg/config"
+	"github.com/shipwright-io/build/pkg/ctxlog"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ReconcileBuild reconciles a Build object
+type ReconcileBuild struct {
+	/* This client, initialized using mgr.Client() above, is a split client
+	   that reads objects from the cache and writes to the apiserver */
+	config                *config.Config
+	client                client.Client
+	scheme                *runtime.Scheme
+	setOwnerReferenceFunc setOwnerReferenceFunc
+}
+
+func NewReconciler(c *config.Config, mgr manager.Manager, ownerRef setOwnerReferenceFunc) reconcile.Reconciler {
+	return &ReconcileBuild{
+		config:                c,
+		client:                mgr.GetClient(),
+		scheme:                mgr.GetScheme(),
+		setOwnerReferenceFunc: ownerRef,
+	}
+}
+
+/* Reconciler finds retentions fields in builds and makes sure the
+   number of corresponding buildruns adhere to these limits */
+func (r *ReconcileBuild) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Set the ctx to be Background, as the top-level context for incoming requests.
+	ctx, cancel := context.WithTimeout(ctx, r.config.CtxTimeOut)
+	defer cancel()
+
+	ctxlog.Debug(ctx, "Start reconciling build-limit-cleanup", namespace, request.Namespace, name, request.Name)
+
+	b := &build.Build{}
+	err := r.client.Get(ctx, request.NamespacedName, b)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			ctxlog.Debug(ctx, "Finish reconciling build-limit-cleanup. Build was not found", namespace, request.Namespace, name, request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
```

Comment:
We are already checking this in the controller. If retention path is set, only then it goes to the reconciler.

Reply:
Understand @PravallikaDindukurthi. What we try to honor here is the asynchonous behavior of all this. The predicate functions in the controller verify if an object should be reconciled. If it says "yes", then the object will added to the reconciliation queue. Eventually it will get picked up and be reconciled. There is a chance that the object might be changed in between. Note that reconcilers always only get the object identity (namespace and name here) and retrieve the object using the client.Get. This will mostly be the same version of the object that the predicate function assessed, but it can be a different one.

Here, we are only talking about a small optimization (omitting the list call for BuildRuns), so one could argue if it is absolutely necessary. In other cases, such race conditions might lead to more severe problems. Because if this it is a general good practice one should always have in mind that in a reconcile function, one cannot rely that the conditions that were evaluated in the predicate function still evaluate to the same result.